docs/contract/confidential-smart-contract.md

```diff
@@ -0,0 +1,264 @@
+   confidential store and
+2. instantiate and call the smart contract without revealing the call arguments.
+
+## PublicCell VS ConfidentialCell
+
+In the [hello world](./hello-world.md) example we used
+`oasis_contract_sdk_storage::cell::PublicCell<T>` to access the key-value store
+of that contract instance. In this case the value was stored unencrypted on the
+blockchain associated with the hash of the key we provided in the constructor.
+
+Cipher supports another primitive
+`oasis_contract_sdk_storage::cell::ConfidentialCell<T>` which allows you to
+access the data of the contract instance confidentially assured by a
+hardware-level encryption.
+
+In contrast to non-confidential calls the hash of the key we passed to the
+constructor, the runtime library also computes a nonce each time the data is
+accessed. This nonce ensures the confidentiality so that the accessed key
+appears different each time from the blockchain observer and even from the node
+operator. It is generated from:
+- the round number,
+- the number of the call inside the current smart contract execution,
+- the number of the smart contract executions until the contract was
+  instantiated.
+
+As an additional safety measure, you will also need to provide the
+instance of the *confidential store* of the contract instance. This is obtained
+by calling `confidential_store()` of the contract's *context* object. If, for
+example, the node operator tries to execute your code in a non-confidential
+environment, this method will panic.
+
+Now, let's look at how a confidential version of the hello world smart contract
+would look like:
+
+```rust
+//! A confidential hello world smart contract.
+extern crate alloc;
+
+use oasis_contract_sdk as sdk;
+use oasis_contract_sdk_storage::cell::ConfidentialCell;
+
+/// All possible errors that can be returned by the contract.
+///
+/// Each error is a triplet of (module, code, message) which allows it to be both easily
+/// human readable and also identifyable programmatically.
+#[derive(Debug, thiserror::Error, sdk::Error)]
+pub enum Error {
+    #[error("bad request")]
+    #[sdk_error(code = 1)]
+    BadRequest,
+}
+
+/// All possible requests that the contract can handle.
+///
+/// This includes both calls and queries.
+#[derive(Clone, Debug, cbor::Encode, cbor::Decode)]
+pub enum Request {
+    #[cbor(rename = "instantiate")]
+    Instantiate { initial_counter: u64 },
+
+    #[cbor(rename = "say_hello")]
+    SayHello { who: String },
+}
+
+/// All possible responses that the contract can return.
+///
+/// This includes both calls and queries.
+#[derive(Clone, Debug, PartialEq, cbor::Encode, cbor::Decode)]
+pub enum Response {
+    #[cbor(rename = "hello")]
+    Hello { greeting: String },
+
+    #[cbor(rename = "empty")]
+    Empty,
+}
+
+/// The contract type.
+pub struct HelloWorld;
+
+/// Storage cell for the counter.
+const COUNTER: ConfidentialCell<u64> = ConfidentialCell::new(b"counter");
```

Comment:
A better example would be one where counters are scoped per-caller, so I can't read your counter value. Or something.

You can use a `ConfidentialMap<Address, u64>` for this.

But of course this is slightly more complicated.

Reply:
I agree, but let's merge the current version first and we'll extend it later. Current TODO in my head
- add `oasis contracts upgrade` to hello-world.md
- add usage of instantiate, upgrade policies to hello-world.md
- add `cargo test` in hello-world.md, mention x86-only for now, try using docker on mac?
- add `ConfidentialMap<K, V>` to confidential-smart-contract.md
- mention that keys are still non-confidential and add `ConfidentialCell<BTreeMap<K,V>>` to make keys confidential too, but for more gas.

Reply:
Sounds good!

> revamp https://github.com/oasisprotocol/oasis-sdk/tree/main/tests/contracts (rename to example-contracts, split to hello-world, confidential-hello-world (do we have abbreviation for "confidential", c10l?), start using code snippets in the docs)

Note that the ones under `tests/contracts` are actually used in E2E tests so they shouldn't be changed. But yes, we can introduce additional "example contracts" as well and start using snippets.